{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "034f9139",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pathlib\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn import metrics\n",
    "from sklearn.metrics import RocCurveDisplay\n",
    "from sklearn.decomposition import NMF\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "from sklearn.mixture import GaussianMixture\n",
    "import os\n",
    "import ot\n",
    "import pickle\n",
    "import argparse\n",
    "import Levenshtein\n",
    "import itertools\n",
    "from helper_functions import *\n",
    "\n",
    "inputdir = \"/media/hieunguyen/HNHD01/raw_data/MRD_GW_v1_20250318/Metadata_Genome-wide_Version1_07.03.25\"\n",
    "\n",
    "PROJECT = \"gs-mrd\"\n",
    "release_version = \"20250319\"\n",
    "\n",
    "##### configurations/paths\n",
    "path_to_main_src = \"/media/hieunguyen/HNSD01/src/gs-mrd/v0.2\"\n",
    "path_to_model_files = f\"{path_to_main_src}/model_files/{release_version}\"\n",
    "path_to_save_features = f\"{path_to_main_src}/model_files/{release_version}/features\"\n",
    "\n",
    "os.system(f\"mkdir -p {path_to_save_features}\")\n",
    "os.system(f\"mkdir -p {path_to_model_files}\")\n",
    "\n",
    "metadata_not_use = pd.read_excel(os.path.join(inputdir, \"Metadata_Genome-wide Version1_07.03.25.xlsx\"))\n",
    "metadata = pd.read_csv(os.path.join(inputdir, \"meta_full.csv\"))\n",
    "\n",
    "##### read features in\n",
    "features = dict()\n",
    "for f in [\"NUCLEOSOME\", \"FLEN\", \"EM\", \"IchorCNA\"]:\n",
    "    features[f] = pd.read_csv(os.path.join(inputdir, f\"{f}.csv\"))\n",
    "features[\"IchorCNA\"].columns = [\"SampleID\", \"ichorCNA\"]\n",
    "\n",
    "##### generate sample list for each class, all samples\n",
    "samplelist = dict()\n",
    "for label in metadata.Cancer.unique():\n",
    "    samplelist[label] = metadata[metadata[\"Cancer\"] == label][\"SampleID\"].to_list()\n",
    "\n",
    "##### generate sample list for each class, train samples only\n",
    "train_samplelist = dict()\n",
    "for label in metadata.Cancer.unique():\n",
    "    train_samplelist[label] = metadata[(metadata[\"Cancer\"] == label) & (metadata[\"Set\"] == \"train\")][\"SampleID\"].to_list()\n",
    "    \n",
    "##### generate sample list for each class, test samples only\n",
    "test_samplelist = dict()\n",
    "for label in metadata.Cancer.unique():\n",
    "    test_samplelist[label] = metadata[(metadata[\"Cancer\"] == label) & (metadata[\"Set\"] == \"test\")][\"SampleID\"].to_list()\n",
    "\n",
    "##### generate sample list for each class, validate samples only\n",
    "validate_samplelist = dict()\n",
    "for label in metadata.Cancer.unique():\n",
    "    validate_samplelist[label] = metadata[(metadata[\"Cancer\"] == label) & (metadata[\"Set\"] == \"validate\")][\"SampleID\"].to_list()\n",
    "\n",
    "##### distance matrix based on edit distance of End motif 4bp\n",
    "nucleotides = ['A', 'C', 'G', 'T']\n",
    "motifs = [''.join(p) for p in itertools.product(nucleotides, repeat=4)]\n",
    "\n",
    "# Initialize an empty distance matrix\n",
    "distance_matrix = pd.DataFrame(index=motifs, columns=motifs)\n",
    "\n",
    "# Compute the Levenshtein distance between each pair of 4-mer motifs\n",
    "for motif1 in motifs:\n",
    "    for motif2 in motifs:\n",
    "        distance_matrix.loc[motif1, motif2] = Levenshtein.distance(motif1, motif2)\n",
    "\n",
    "# Convert the distance matrix to integer type\n",
    "M_EM = distance_matrix.to_numpy().copy()\n",
    "M_EM /= M_EM.max() * 0.1\n",
    "\n",
    "final_features = dict() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "0a503a35",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['Healthy', 'Lung', 'Breast', 'HCC', 'Ovarian', 'CRC', 'Gastric'])"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samplelist.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "434f89da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SampleID</th>\n",
       "      <th>Label</th>\n",
       "      <th>Set</th>\n",
       "      <th>Cancer</th>\n",
       "      <th>Run</th>\n",
       "      <th>Kit</th>\n",
       "      <th>Team run</th>\n",
       "      <th>System</th>\n",
       "      <th>Group (Early/Met)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZK0AAAI82NB</td>\n",
       "      <td>Neg</td>\n",
       "      <td>train</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>R5484</td>\n",
       "      <td>No UID</td>\n",
       "      <td>ECD</td>\n",
       "      <td>MGI</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ZK0AAAH81NB</td>\n",
       "      <td>Neg</td>\n",
       "      <td>train</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>R5451</td>\n",
       "      <td>No UID</td>\n",
       "      <td>ECD</td>\n",
       "      <td>MGI</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZK0CAAA56NB</td>\n",
       "      <td>Neg</td>\n",
       "      <td>train</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>R5170</td>\n",
       "      <td>No UID</td>\n",
       "      <td>ECD</td>\n",
       "      <td>MGI</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZK0AAAA16NB</td>\n",
       "      <td>Neg</td>\n",
       "      <td>train</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>R5063</td>\n",
       "      <td>No UID</td>\n",
       "      <td>ECD</td>\n",
       "      <td>MGI</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ZK0CAAA72NB</td>\n",
       "      <td>Neg</td>\n",
       "      <td>train</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>R5184</td>\n",
       "      <td>No UID</td>\n",
       "      <td>ECD</td>\n",
       "      <td>MGI</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1325</th>\n",
       "      <td>ZMC039NB</td>\n",
       "      <td>Pos</td>\n",
       "      <td>validate</td>\n",
       "      <td>CRC</td>\n",
       "      <td>R5169</td>\n",
       "      <td>No UID</td>\n",
       "      <td>ECD</td>\n",
       "      <td>MGI</td>\n",
       "      <td>Early</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1326</th>\n",
       "      <td>ZYCAB05NB</td>\n",
       "      <td>Pos</td>\n",
       "      <td>validate</td>\n",
       "      <td>Ovarian</td>\n",
       "      <td>R5167</td>\n",
       "      <td>No UID</td>\n",
       "      <td>ECD</td>\n",
       "      <td>MGI</td>\n",
       "      <td>Early</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1327</th>\n",
       "      <td>ZMB159NB</td>\n",
       "      <td>Pos</td>\n",
       "      <td>validate</td>\n",
       "      <td>Breast</td>\n",
       "      <td>R5170</td>\n",
       "      <td>No UID</td>\n",
       "      <td>ECD</td>\n",
       "      <td>MGI</td>\n",
       "      <td>Early</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1328</th>\n",
       "      <td>ZMB059NB</td>\n",
       "      <td>Pos</td>\n",
       "      <td>validate</td>\n",
       "      <td>Breast</td>\n",
       "      <td>R5083</td>\n",
       "      <td>No UID</td>\n",
       "      <td>ECD</td>\n",
       "      <td>MGI</td>\n",
       "      <td>Early</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1329</th>\n",
       "      <td>ZMG093</td>\n",
       "      <td>Pos</td>\n",
       "      <td>validate</td>\n",
       "      <td>Gastric</td>\n",
       "      <td>R4810</td>\n",
       "      <td>UID</td>\n",
       "      <td>MRD</td>\n",
       "      <td>N2K</td>\n",
       "      <td>Early</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1330 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         SampleID Label       Set   Cancer    Run     Kit Team run System  \\\n",
       "0     ZK0AAAI82NB   Neg     train  Healthy  R5484  No UID      ECD    MGI   \n",
       "1     ZK0AAAH81NB   Neg     train  Healthy  R5451  No UID      ECD    MGI   \n",
       "2     ZK0CAAA56NB   Neg     train  Healthy  R5170  No UID      ECD    MGI   \n",
       "3     ZK0AAAA16NB   Neg     train  Healthy  R5063  No UID      ECD    MGI   \n",
       "4     ZK0CAAA72NB   Neg     train  Healthy  R5184  No UID      ECD    MGI   \n",
       "...           ...   ...       ...      ...    ...     ...      ...    ...   \n",
       "1325     ZMC039NB   Pos  validate      CRC  R5169  No UID      ECD    MGI   \n",
       "1326    ZYCAB05NB   Pos  validate  Ovarian  R5167  No UID      ECD    MGI   \n",
       "1327     ZMB159NB   Pos  validate   Breast  R5170  No UID      ECD    MGI   \n",
       "1328     ZMB059NB   Pos  validate   Breast  R5083  No UID      ECD    MGI   \n",
       "1329       ZMG093   Pos  validate  Gastric  R4810     UID      MRD    N2K   \n",
       "\n",
       "     Group (Early/Met)  \n",
       "0                  NaN  \n",
       "1                  NaN  \n",
       "2                  NaN  \n",
       "3                  NaN  \n",
       "4                  NaN  \n",
       "...                ...  \n",
       "1325             Early  \n",
       "1326             Early  \n",
       "1327             Early  \n",
       "1328             Early  \n",
       "1329             Early  \n",
       "\n",
       "[1330 rows x 9 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mrd",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
